# Assignment 1

**Group 22**

A. Siganos ( 1283871 )     
J. Gómez Robles ( 1286552 )

In [37]:
import numpy as np
np.random.seed(13) #TODO Check if this is used for sgd
import keras.backend as K
from keras.models import Sequential
from keras.layers import Dense, Embedding, Reshape, Lambda
from keras.utils import np_utils
from keras.utils.data_utils import get_file
from keras.preprocessing.text import Tokenizer
from keras.utils.vis_utils import model_to_dot
from keras.preprocessing import sequence
from gensim.models import KeyedVectors
from sklearn.manifold import TSNE
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.neighbors import NearestNeighbors as nn
from matplotlib import pylab
from __future__ import division

import time
# Only for the word2vec parsing of last exercise (it was part of the template, not our implementation)
#import gensim


In [27]:
# DO NOT Modify the lines in this cell
path = 'alice.txt'
corpus = open(path).readlines()[0:700]

corpus = [sentence for sentence in corpus if sentence.count(" ") >= 2]

tokenizer = Tokenizer(filters='!"#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n'+"'")
tokenizer.fit_on_texts(corpus)
corpus = tokenizer.texts_to_sequences(corpus)
nb_samples = sum(len(s) for s in corpus)
V = len(tokenizer.word_index) + 1

# Is this something they need to change?
dim = 100
window_size = 2 #use this window size for Skipgram, CBOW, and the model with the additional hidden layer
window_size_corpus = 4 #use this window size for the co-occurrence matrix

## Question 1

### Co-occurrence Matrix
Use the provided code to load the "Alice in Wonderland" text document. 
1. Implement the word-word co-occurrence matrix for “Alice in Wonderland”
2. Normalize the words such that every value lies within a range of 0 and 1
3. Compute the cosine distance between the given words:
    - Alice 
    - Dinah
    - Rabbit
4. List the 5 closest words to 'Alice'. Discuss the results.
5. Discuss what the main drawbacks are of a term-term co-occurence matrix solutions?


In [28]:
################################################################################
############################## 0. Util functions ###############################
################################################################################

# Run this block first
def getKeyFromValue(dictionary, target):
    return list(dictionary.keys())[list(dictionary.values()).index(target)]


In [29]:
################################################################################
################# 1. Implement word-word co-ocurrence matrix ###################
################################################################################

# Create my co-ocurrence matrix, initially 0 (VxV size)
# How to access this matrix:
#   Each column is an (index - 1) (column 0 is word in index 1; column 1 is word in index 2; ...)
wcoMatrix = np.zeros([V, V]) 

# Using my window_size_corpus to define my context scope
scope = window_size_corpus

# TODO: Do not make that many iterations
# Greedy approach first (to be able to compare the optimization)

# For each line in the corpus. Note that they preserve the order, even when they are indexes now.
for s in corpus:
    # For each word
    for current_index in range(0, len(s)):
        current_value = s[current_index] # Get the 'word'
        # From left to right
        for neighbor_index in range(current_index - scope, current_index + scope + 1):
            # Never out of boundaries and never the same index
            if ( neighbor_index >= 0 and neighbor_index < len(s) ) and ( neighbor_index != current_index ):
                # Get my neighbor 'word'
                neighbor = s[neighbor_index] 
                # Never myself and myself (keeping diagonal to 0)
                if current_value != neighbor:
                    # Update the ocurrence
                    #wcoMatrix[current_value - 1, neighbor - 1] += 1
                    wcoMatrix[current_value, neighbor] += 1
                    
################################################################################
################################# 2. Normalize #################################
################################################################################
wcoMatrix = wcoMatrix / wcoMatrix.max()
wcoMatrix

print("Sumary of the word-word co-ocurrence matrix")
print(" - Shape", wcoMatrix.shape)
print(" - Min value", wcoMatrix.min())
print(" - Max value", wcoMatrix.max())


Sumary of the word-word co-ocurrence matrix
 - Shape (1183, 1183)
 - Min value 0.0
 - Max value 1.0


In [30]:
################################################################################
########### 3. Compute cosine similarity to Alice, Dinah and Rabbit ############
################################################################################

words_to_compare = ["Alice", "Dinah", "Rabbit"]

# Get my index
full_index = tokenizer.word_index

# Iterate on all possible combinations
for w1 in words_to_compare:
    w1 = w1.lower() 
    x = full_index[w1]
    for w2 in words_to_compare:
        w2 = w2.lower()
        if w1 != w2:
            y = full_index[w2]
            X = wcoMatrix[x, :].reshape((1, V))
            Y = wcoMatrix[y, :].reshape((1, V))
            print("cosine_similarity(%s, %s)=%s" % (w1, w2, cosine_similarity(X, Y)))
            

cosine_similarity(alice, dinah)=[[0.39360011]]
cosine_similarity(alice, rabbit)=[[0.47890931]]
cosine_similarity(dinah, alice)=[[0.39360011]]
cosine_similarity(dinah, rabbit)=[[0.29862324]]
cosine_similarity(rabbit, alice)=[[0.47890931]]
cosine_similarity(rabbit, dinah)=[[0.29862324]]


In [31]:
################################################################################
################## 4a. List the 5 closest words to 'Alice'. ####################
################################################################################

################################
### We use Nearest Neighbors ###
################################

# It was not importing
from sklearn.neighbors import NearestNeighbors

# Define our variables
targetWord = "Alice".lower()
X = wcoMatrix
N_NEIGHBORS = 5
full_index = tokenizer.word_index

# Perfrom the NearestNeibhbors algorithm
# TODO: Tune parameters
nbrs = NearestNeighbors(n_neighbors = N_NEIGHBORS).fit(X)

# Get the index corresponding to the target word
aliceMatrixIndex = full_index[targetWord]

# Get results from nearest neighbors
distances, indices = nbrs.kneighbors()
aliceIndices = indices[aliceMatrixIndex, :]
aliceDistances = distances[aliceMatrixIndex, :]

# Output result
print("Nearest neighbors (co-ocurrence matrix based) to %s (index %s):" % (targetWord, aliceMatrixIndex))
for i in range(0, N_NEIGHBORS):
    print( "%s [index %s] with distance: %s" % (getKeyFromValue(full_index, aliceIndices[i]), aliceIndices[i], aliceDistances[i]) )

# TODO: Think about what we are computing. 
# With first approach, a neighbor would be a word with a similar vector in the co-ocurrence matrix, 
# not a word within the same context (window_size). Although words in the similar window_size should have a
# similar vector, isn't?

Nearest neighbors (co-ocurrence matrix based) to alice (index 11):
her [index 14] with distance: 0.5616433150108668
that [index 13] with distance: 0.5692975698101006
herself [index 41] with distance: 0.5700573426956416
for [index 18] with distance: 0.5811858638252018
on [index 21] with distance: 0.5837124477057559


**4b. Discuss the results**

We observe that the most similar words are possesives and prepositions. Although these results intuitively make some sense (Alice is the main character of the book and consequently, describing her behaviour is expected), no idiomatic relations can be achieved by means of the simple word-word co-ocurrence matrix.

Furthermore, the minimal distance (cosine similarity normalized) found for alice is of 0.5616, which suggests a poor result - one would be interested on a similarity close to 0.


**5. Discuss what the main drawbacks are of a term-term co-occurence matrix solutions? **

TODO. Add some ideas!

thanos: It requires large memory to store the co-occurrence matrix. 
Jorge's ideas: sparse matrix VxV size (computation time increases).



In [32]:
#Save your all the vector representations of your word embeddings in this way
#Change when necessary the sizes of the vocabulary/embedding dimension

f = open('vectors_co_occurrence.txt',"w")
f.write(" ".join([str(V-1),str(V-1)]))
f.write("\n")

#vectors = your word co-occurrence matrix
vectors = wcoMatrix
for word, i in tokenizer.word_index.items():    
    f.write(word)
    f.write(" ")
    f.write(" ".join(map(str, list(vectors[i,:]))))
    f.write("\n")
f.close()


In [33]:
#reopen your file as follows
# TODO: Check! It is not reloading the file D:
# co_occurrence = KeyedVectors.load_word2vec_format('./vectors_co_occurrence.txt', binary=False)


## Question 2

### Word embeddings
Build embeddings with a keras implementation where the embedding vector is of length 50, 150 and 300. Use the Alice in Wonderland text book for training.
1. Using the CBOW model
2. Using Skipgram model
3. Add extra hidden dense layer to CBow and Skipgram implementations. Choose an activation function for that layer and justify your answer.
4. Analyze the four different word embeddings
    - Implement your own function to perform the analogy task with. Do not use existing libraries for this task such as Gensim. Your function should be able to answer whether an anaology as in the example given in the pdf-file is true.
    - Compare the performance on the analogy task between the word embeddings that you have trained in 2.1, 2.2 and 2.3.  
    - Visualize your results and interpret your results
5. Use the word co-occurence matrix from Question 1. Compare the performance on the analogy task with the performance of your trained word embeddings.  
6. Discuss:
    - What are the main advantages of CBOW and Skipgram?
    - What is the advantage of negative sampling?
    - What are the main drawbacks of CBOW and Skipgram?
7. Load pre-trained embeddings on large corpuses (see the pdf file). You only have to consider the word embeddings with an embedding size of 300
    - Compare performance on the analogy task with your own trained embeddings from "Alice in Wonderland". You can limit yourself to the vocabulary of Alice in Wonderland. Visualize the pre-trained word embeddings and compare these with the results of your own trained word embeddings. 


# CBOW

In [34]:
#prepare data for cbow
def generate_data_cbow(corpus, window_size, V):
    all_in = [] 
    all_out = []
    maxlen = window_size*2
    for words in corpus:
        L = len(words)
        for index, word in enumerate(words):
            contexts = []
            labels   = []            
            s = index - window_size
            e = index + window_size + 1
            
            contexts.append([words[i] for i in range(s, e) if 0 <= i < L and i != index])
            labels.append(word)
            if contexts != []:
                all_in.append(sequence.pad_sequences(contexts, maxlen = maxlen))
                all_out.append(np_utils.to_categorical(labels, V))
    return (all_in,all_out) # return the values

            
            
# def generate_data_skipgram(corpus, window_size, V):
#     all_in = [] 
#     all_out = []
#     for sentence in corpus: # For, sentence in corpus
#         L = len(sentence) # Limit for the iteration (per sentence)

#         # For each pair in the index
#         for index, word in enumerate(sentence):
#             # Create X and Y
#             in_words = [] # list of lists
#             labels = [] # list of integers

#             # Iterate over all the sentence, for each word in the line. I.e, for each word:
#             for i in range(index - window_size, index + window_size + 1):
#                 # If between the limits and not myself
#                 if i != index and (0 <= i < L):
#                     # This means: Associating the words in my neighbor (context) 
#                     in_words.append([word]) # Current word (as a list) is associated to...
#                     labels.append(sentence[i]) # ... this word (as a simple integer).
                    
#             # Once I have saved the context for the current word...
#             # ... if there is something in the arrays
#             if in_words != []:
#                 # Save the information to the all_in/out arrays
#                 all_in.append(np.array(in_words,dtype=np.int32)) # X in integer format
#                 # Set Y to categorical values. If my output is ([6, 26]). Then I create
#                 # a vector from 0 to V and mark 6 and 26 as 1
#                 all_out.append(np_utils.to_categorical(labels, V))
#     return (all_in,all_out) # return the values



            

In [59]:
#create CBOW model

def getCbowModel(dim = 50, extra_layer = False, activation_extra_layer = 'softmax'):

    # The Sequential model is a linear stack of layers.
    cbow = Sequential()
    cbow.add(Embedding(input_dim = V, output_dim = dim, input_length = window_size*2))

    # TODO!!! 
    if extra_layer :
        cbow.add(Dense(input_dim = V, units=dim, kernel_initializer='uniform', activation = activation_extra_layer))
        # input_dim=V input ,units=dim output
    cbow.add( Lambda(lambda x: K.mean(x, axis = 1), output_shape = (dim,)) )
    cbow.add( Dense(V, activation = 'softmax') )

    # define loss function for Cbow
    cbow.compile(loss='categorical_crossentropy', optimizer = 'adadelta')

    return cbow


In [39]:
#train model

################################################################################
############################ 1. Using CBOW model ###############################
################################################################################

dims = [50, 150, 300]
iterations = 10
listResX = []
listResY = []

for dimension in dims:
    print("Running CBOW (dim = %s)..."% dimension)
    cbow = getCbowModel(dimension, False)
    X, Y = generate_data_cbow(corpus, window_size, V)
    start_time=time.time()
    for ite in range(10):
        loss = 0.
        for x, y in zip(X, Y):
            loss += cbow.train_on_batch(x, y)
        print(ite, loss)

    end_time = time.time()
    
    print("--- %s seconds ---" % (end_time - start_time))
    listResX.append(X)
    listResY.append(Y)


Running CBOW (dim = 50)...
0 41690.76107144356
1 39117.85015499592
2 38920.646042943
3 38817.39956218004
4 38749.09197306633
5 38693.82797335088
6 38635.66837017238
7 38583.93498829007
8 38544.42379313707
9 38519.23661490157
--- 173.09521794319153 seconds ---
Running CBOW (dim = 150)...
0 41625.12459421158
1 38779.02264201641
2 38344.01991537213
3 38101.7295871675
4 37894.86858777702
5 37698.67937242985
6 37520.287830296904
7 37375.063762228936
8 37265.025551433675
9 37175.7061145301
--- 326.6905765533447 seconds ---
Running CBOW (dim = 300)...
0 41555.244079589844
1 38479.914032936096
2 37939.93286758661
3 37621.721390232444
4 37351.13504803181
5 37100.079870630056
6 36880.70205508545
7 36706.02954676375
8 36560.097589663696
9 36428.92054900038
--- 1484.4708473682404 seconds ---


# Skipgram

In [43]:
#prepare data for Skipgram

# This function creates a list where _each word_ in the text file (with non-empty content)
# is encoded into two arrays:
#   - x: an array of the size of all his windows_size neighbors with the value of the word (i.e, repeated values)
#   - y: an array of the same size as x pointing to each neighbor. Finally encoded in a |V|-list as one-hot enc
def generate_data_skipgram(corpus, window_size, V):
    all_in = [] 
    all_out = []
    for sentence in corpus: # For, sentence in corpus
        L = len(sentence) # Limit for the iteration (per sentence)

        # For each pair in the index
        for index, word in enumerate(sentence):
            # Create X and Y
            in_words = [] # list of lists
            labels = [] # list of integers

            # Iterate over all the sentence, for each word in the line. I.e, for each word:
            for i in range(index - window_size, index + window_size + 1):
                # If between the limits and not myself
                if i != index and (0 <= i < L):
                    # This means: Associating the words in my neighbor (context) 
                    in_words.append([word]) # Current word (as a list) is associated to...
                    labels.append(sentence[i]) # ... this word (as a simple integer).
                    
            # Once I have saved the context for the current word...
            # ... if there is something in the arrays
            if in_words != []:
                # Save the information to the all_in/out arrays
                all_in.append(np.array(in_words,dtype=np.int32)) # X in integer format
                # Set Y to categorical values. If my output is ([6, 26]). Then I create
                # a vector from 0 to V and mark 6 and 26 as 1
                all_out.append(np_utils.to_categorical(labels, V))
    return (all_in,all_out) # return the values

# Prepare data for skipgram
#X, Y = generate_data_skipgram(corpus, window_size, V)


In [44]:
# #save the preprocessed data of Skipgram
# f = open('data_skipgram.txt' ,'w')

# for input, outcome in zip(X,Y):
#     input = np.concatenate(input)
#     f.write(" ".join(map(str, list(input))))
#     f.write(",")
#     outcome = np.concatenate(outcome)
#     f.write(" ".join(map(str,list(outcome))))
#     f.write("\n")
# f.close()


In [45]:
# #load the preprocessed Skipgram data
# def generate_data_skipgram_from_file():
#     f = open('data_skipgram.txt' ,'r')
#     for row in f:
#         inputs,outputs = row.split(",")
#         inputs = np.fromstring(inputs, dtype=int, sep=' ')
#         inputs = np.asarray(np.split(inputs, len(inputs)))
#         outputs = np.fromstring(outputs, dtype=float, sep=' ')
#         outputs = np.asarray(np.split(outputs, len(inputs)))
#         yield (inputs,outputs)
        

In [46]:
# create Skipgram model

def getSkipgramModel(dim = 50, extra_layer = False, activation_extra_layer = 'softmax'):

    # The Sequential model is a linear stack of layers.
    skipgram = Sequential()
    
    # First layer (Projection layer. Embedding for wt)
    # 
    # Parameters: 
    #   - input_dim: V (From the book: 1xV (input layer))
    #   - embeddings_initializer = Glorot uniform. Is it for the uniform distribution of the noises?
    #   - input_length: 1. It is needed for the dense layer (https://keras.io/layers/embeddings/#embedding)
    # Input: One-hot ecoding of word wt in the 1xV space
    # Output: vj in the dimension space 1xd
    skipgram.add( Embedding(input_dim = V, output_dim = dim, embeddings_initializer = 'glorot_uniform', input_length = 1) )

    # Intermediate layer. Can be seen as a helper only.
    skipgram.add( Reshape((dim, )) )

    # Second Layer (Output layer. Probabilities of context words)
    # From documentation, Dense implements the operation: 
    #      output = activation(dot(input, kernel) + bias) 
    #   where:
    #      activation is the element-wise activation function passed as the activation argument
    #      kernel is a weights matrix created by the layer
    #      and bias is a bias vector created by the layer (only applicable if use_bias is True). _WE DO NOT SET THIS_.
    #
    # Hence, we are computing ck*vj
    #
    # Parameters:
    #   - input_dim: dim. From embedding layer, dx1
    #   - units: V. The output dimension, From the book 1xV
    #   - kernel_initializer: uniform (?)
    #   - activation: softmax (as described in the book)
    if extra_layer :
        skipgram.add(Dense(input_dim=dim, units=V, kernel_initializer='uniform', activation= activation_extra_layer))

    skipgram.add( Dense(input_dim=dim, units=V, kernel_initializer='uniform', activation='softmax') )
    
    # Extra layer ()
    #
    # Justification: 
    #
    #
#     if extra_layer :
#         skipgram.add(Dense(input_dim=dim, units=V, kernel_initializer='uniform', activation= activation_extra_layer))
    
    # define loss function for Skipgram
    skipgram.compile(loss='categorical_crossentropy', optimizer='adadelta')

    return skipgram

In [17]:
################################################################################
########################## 2. Using Skipgram model #############################
################################################################################

dims = [50, 150, 300]
iterations = 10
listResX = []
listResY = []

for dimension in dims:
    print("Running Skipgram (dim = %s)..."% dimension)
    skipgram = getSkipgramModel(dimension, False)
    #print(skipgram.summary())
    X, Y = generate_data_skipgram(corpus, window_size, V)

    start_time = time.time()
    for ite in range(iterations):
        loss = 0.
        for x, y in zip(X,Y):
            loss += skipgram.train_on_batch(x, y)=
        print(ite, loss)
    end_time = time.time()
    
    print("--- %s seconds ---" % (end_time - start_time))
    listResX.append(X)
    listResY.append(Y)


Running Skipgram (dim = 50)...
0 41271.35240674019
1 39126.84181976318
2 39288.22808551788
3 39363.864663124084
4 39432.38656759262
5 39511.96853172779
6 39608.38287115097
7 39723.89285326004
8 39852.42831707001
9 39987.01795423031
--- 178.2564082145691 seconds ---
Running Skipgram (dim = 150)...
0 41222.66726255417
1 38923.405826091766
2 38988.34154701233
3 39005.65850484371
4 39038.850784897804
5 39093.706859230995
6 39156.13746762276
7 39216.62943792343
8 39267.45716762543
9 39305.05479335785
--- 347.22299909591675 seconds ---
Running Skipgram (dim = 300)...
0 41159.68823719025
1 38730.84329485893
2 38713.535183906555
3 38676.76882696152
4 38648.035522818565
5 38631.88415277004
6 38611.77810204029
7 38578.923778772354
8 38533.39666950703
9 38484.56395339966
--- 741.4052288532257 seconds ---


# CBOW additional hidden layer

In [60]:
################################################################################
################### 3. Using Skipgram model (extra layer) ######################
################################################################################

#dims = [50, 150, 300]
dims = [300]
iterations = 2
listResX = []
listResY = []
activation_rules = ["softmax", "elu", "selu", "softplus", "softsign", "relu", "tanh", "sigmoid", "hard_sigmoid", "linear"]
for dimension in dims:
    for activation_rule in activation_rules:
        print("Running Skipgram (dim = %s) with activation rule '%s'..."% (dimension, activation_rule))
#         try:
        # set extra_layer = true
        cbow = getCbowModel(dimension, True, activation_rule)
        X, Y = generate_data_cbow(corpus, window_size, V)
        
        start_time = time.time()
        for ite in range(iterations):
            loss = 0.
            for x, y in zip(X,Y):
                loss += cbow.train_on_batch(x, y)
            print(ite, loss)
        end_time = time.time()

        print("--- %s seconds ---" % (end_time - start_time))
        listResX.append(X)
        listResY.append(Y)
        


#         except:
#             print("Error with", activation_rule)
        

Running Skipgram (dim = 300) with activation rule 'softmax'...
0 41637.214460134506
1 39596.0313744545
--- 200.4243266582489 seconds ---
Running Skipgram (dim = 300) with activation rule 'elu'...
0 39377.03297561407
1 37039.42694170773
--- 202.7668013572693 seconds ---
Running Skipgram (dim = 300) with activation rule 'selu'...
0 39139.040005236864
1 36735.84380609915
--- 205.15451550483704 seconds ---
Running Skipgram (dim = 300) with activation rule 'softplus'...
0 40226.599430680275
1 39699.8060836792
--- 201.13255524635315 seconds ---
Running Skipgram (dim = 300) with activation rule 'softsign'...
0 39696.88131862879
1 37257.71322147548
--- 202.3088939189911 seconds ---
Running Skipgram (dim = 300) with activation rule 'relu'...
0 39568.10435950756
1 38047.06096044183
--- 210.33026313781738 seconds ---
Running Skipgram (dim = 300) with activation rule 'tanh'...
0 39412.67246413231
1 37054.29612833634
--- 218.752925157547 seconds ---
Running Skipgram (dim = 300) with activation rule

# Skipgram additional hidden layer

In [47]:
################################################################################
################### 3. Using Skipgram model (extra layer) ######################
################################################################################

#dims = [50, 150, 300]
dims = [50]
iterations = 2
listResX = []
listResY = []
activation_rules = ["softmax", "elu", "selu", "softplus", "softsign", "relu", "tanh", "sigmoid", "hard_sigmoid", "linear"]
for dimension in dims:
    for activation_rule in activation_rules:
        print("Running Skipgram (dim = %s) with activation rule '%s'..."% (dimension, activation_rule))
        try:
            # set extra_layer = true
            skipgram = getSkipgramModel(dimension, True, activation_rule)
            X, Y = generate_data_skipgram(corpus, window_size, V)

            start_time = time.time()
            for ite in range(iterations):
                loss = 0.
                for x, y in zip(X,Y):
                    loss += skipgram.train_on_batch(x, y)
                print(ite, loss)
            end_time = time.time()

            print("--- %s seconds ---" % (end_time - start_time))
            listResX.append(X)
            listResY.append(Y)

        except:
            print("Error with", activation_rule)
        

Running Skipgram (dim = 50) with activation rule 'softmax'...
Error with softmax
Running Skipgram (dim = 50) with activation rule 'elu'...
Error with elu
Running Skipgram (dim = 50) with activation rule 'selu'...
Error with selu
Running Skipgram (dim = 50) with activation rule 'softplus'...
Error with softplus
Running Skipgram (dim = 50) with activation rule 'softsign'...
Error with softsign
Running Skipgram (dim = 50) with activation rule 'relu'...
Error with relu
Running Skipgram (dim = 50) with activation rule 'tanh'...
Error with tanh
Running Skipgram (dim = 50) with activation rule 'sigmoid'...
Error with sigmoid
Running Skipgram (dim = 50) with activation rule 'hard_sigmoid'...
Error with hard_sigmoid
Running Skipgram (dim = 50) with activation rule 'linear'...
Error with linear


In [22]:
################################################################################
####### 4a. Implement your own function to perform the analogy task with #######
################################################################################

embeddingWcoMatrix = "wcoMatrix"
# objA is to objB as objC is a objD
# we assume objs are strings
def analogyFunction(objA, objB, objC, objD, embeddingType = embeddingWcoMatrix, matrix = wcoMatrix):
    indexA = full_index[objA]
    indexB = full_index[objB]
    indexC = full_index[objC]
    indexD = full_index[objD]
    
    if(embeddingType == embeddingWcoMatrix):
        vecA = np.asarray(matrix[indexA,])
        vecB = np.asarray(matrix[indexB,])
        vecC = np.asarray(matrix[indexC,])
        vecD = np.asarray(matrix[indexD,])
    
    operationResult = vecA - vecB + vecD
    return ( np.array_equal(vecC,operationResult) ) # Not sure if this works

def makeAnalogyOnFile(embeddingType = embeddingWcoMatrix, matrix = wcoMatrix):
    # Read file
    content = []
    with open("analogy_alice.txt") as f:
        content = f.readlines()

    # Remove whitespace characters like '\n' at the end of each line
    content = [x.strip() for x in content] 
    allResults = []
    for analogy in content:
        words = analogy.split(" ")
        if len(words) == 4:
            try:
                result = analogyFunction(words[0], words[1], words[2], words[3])
                allResults.append(result)
                print("Testing '%s': %s" % (analogy, result))
            except KeyError:
                print("Missing words in the index ")
    return allResults

Comparison performance:

In [23]:
################################################################################
########### 4b. Compare the performance on the analogy task between ############
########### the word embeddings you have trained in 2.1, 2.2 & 2.3  ############
################################################################################


# 4. Analyze the four different word embeddings
#    - Compare the performance on the analogy task between the word embeddings that you have trained in 2.1, 2.2 and 2.3.  
#    - Visualize your results and interpret your results
# 5. Use the word co-occurence matrix from Question 1. Compare the performance on the analogy task with the performance of your trained word embeddings.  


analogyResultsWcoMatrix = makeAnalogyOnFile()
print("Score for w-w co-ocurrence matrix embedding: %s" % sum(analogyResultsWcoMatrix))

Missing words in the index 
Testing 'sudden suddenly usual usually': False
Testing 'bad worse good better': False
Testing 'go going look looking': False
Testing 'he she his her': False
Testing 'brother sister his her': False
Testing 'listen listening look looking': False
Testing 'saying said thinking thought': False
Testing 'bird birds cat cats': False
Testing 'good better old older': False
Testing 'good better quick quicker': False
Testing 'large largest good best': False
Missing words in the index 
Testing 'falling fell knowing knew': False
Testing 'walk walking think thinking': False
Testing 'child children cat cats': False
Testing 'dog dogs eye eyes': False
Testing 'hand hands rat rats': False
Testing 'eat eats find finds': False
Testing 'find finds say says': False
Testing 'old older good better': False
Testing 'large larger quick quicker': False
Testing 'go going listen listening': False
Testing 'run running walk walking': False
Testing 'run running think thinking': False
Missing

In [24]:
# Visualize results

# pylint: disable=missing-docstring
# Function to draw visualization of distance between embeddings.
def plot_with_labels(low_dim_embs, labels, filename):
  assert low_dim_embs.shape[0] >= len(labels), 'More labels than embeddings'
  plt.figure(figsize=(18, 18))  # in inches
  for i, label in enumerate(labels):
    x, y = low_dim_embs[i, :]
    plt.scatter(x, y)
    plt.annotate(label,
                 xy=(x, y),
                 xytext=(5, 2),
                 textcoords='offset points',
                 ha='right',
                 va='bottom')

  plt.savefig(filename)

try:
  # pylint: disable=g-import-not-at-top
  from sklearn.manifold import TSNE
  import matplotlib.pyplot as plt

  tsne = TSNE(perplexity=30, n_components=2, init='pca', n_iter=5000, method='exact')
  plot_only = 500
  low_dim_embs = tsne.fit_transform(final_embeddings[:plot_only, :])
  labels = [reverse_dictionary[i] for i in xrange(plot_only)]
  plot_with_labels(low_dim_embs, labels, os.path.join(gettempdir(), 'tsne.png'))

except ImportError as ex:
  print('Please install sklearn, matplotlib, and scipy to show embeddings.')
  print(ex)

NameError: name 'final_embeddings' is not defined

In [ ]:
################################################################################
############# 5. Use the word co-occurence matrix from Question 1. #############
################################################################################

# Compare the performance on the analogy task with the performance of your trained word embeddings.  
analogyResultsWcoMatrix = makeAnalogyOnFile()

print("\nScore for w-w co-ocurrence matrix embedding: %s" % sum(analogyResultsWcoMatrix))

Interpretation results of the visualization

Compare the results of the trained word embeddings with the word-word co-occurrence matrix

**6. Discuss**


#### What are the main advantages of CBOW and Skipgram?


General pros:
It preserves the semantic relationship between words. i.e man and woman tend to be closer than man and apple.
It uses factorization which is a well-defined problem and can be efficiently solved.
It has to be computed once and can be used anytime once computed. 

CBOW:
-It is low on memory. It does not need to have huge RAM requirements like that of co-occurrence matrix where it needs to store -three huge matrices.
-CBOW is faster to train and works well with frequent words

skipgram:
-Skip-gram model can capture two semantics for a single word. i.e it will have two vector representations of Apple. One for the company and other for the fruit.
-Skip-Gram is more efficient with small training data
-infrequent words are well presented

-------------Difference-------------

We know that CBOW is learning to predict the word based on the context. We can simply "translate" that as maximizing the probability of the target word by looking at the context. This is why CBOW works better with frequent words but not the same happens for rare words. For example, given the context "yesterday was really [...] day" CBOW model will tell us that most probably the word is "beautiful" or "nice", which are quite usually used words. Words like "delightful" on the other hand will get much less attention of the model, because it is designed to predict the most probable word. This word will be smoothed over a lot of examples with more frequent words.

On the other hand, the skip-gram is designed to predict the context based on the target word. Given the word "delightful" it must study it and tell us, that there is huge probability, the context is "yesterday was really [...] day", or some other relevant context. With skip-gram the word delightful will not try to compete with word beautiful but instead, delightful+context pairs will be treated as new observations


#### What is the advantage of negative sampling?
Negative sampling tends to work better for frequent words and better with low dimensional vectors. 

To optimize performance for appropriate data (frequent words,small data sets) we would use CBOW with negative sampling 


Negative sampling is a way to sample the training data, similar to stochastic gradient descent, but the key is you look for negative training examples. Intuitively, it trains based on sampling places it might have expected a word, but didn't find one, which is faster than training an entire corpus every iteration and makes sense for common words.

Negative sampling is developed based on noise contrastive estimation and randomly samples the words not in the context to distinguish the observed data from the artificially generated random noise.

#### What are the main drawbacks of CBOW and Skipgram?

-CBOW takes the average of the context of a word (as seen above in calculation of hidden activation). For example, Apple can be both a fruit and a company but CBOW takes an average of both the contexts and places it in between a cluster for fruits and companies.
-Training a CBOW from scratch can take forever if not properly optimized

Skipgram takes a substantially larger amount of time since rather than averaging the context words, each context word is used as a pairwise training example. That is, in place of one CBOW example such as [predict 'ate' from average('The', 'cat', 'the', 'mouse')], the network is presented with four skip-gram examples [predict 'ate' from 'The'], [predict 'ate' from 'cat'], [predict 'ate' from 'the'], [predict 'ate' from 'mouse']




7 Load pre-trained embeddings on large corpuses (see the pdf file). You only have to consider the word embeddings with an embedding size of 300

    - Compare performance on the analogy task with your own trained embeddings from "Alice in Wonderland". You can limit yourself to the vocabulary of Alice in Wonderland. Visualize the pre-trained word embeddings and compare these with the results of your own trained word embeddings. 


In [ ]:
#load pretrained word embeddings of word2vec
path_word2vec = "./GoogleNews-vectors-negative300.bin"
word2vec = KeyedVectors.load_word2vec_format(path_word2vec, binary=True)


In [ ]:
#load pretraind word embeddings of Glove

#path = "./glove.6B/glove.6B.300d_converted.txt"
path = "./glove.6B/glove.6B.300d.txt"

#convert GloVe into word2vec format
gensim.scripts.glove2word2vec.get_glove_info(path)
gensim.scripts.glove2word2vec.glove2word2vec(path, "glove_converted.txt")

glove = KeyedVectors.load_word2vec_format(path, binary=False)


In [ ]:
#Visualize the pre-trained word embeddings


Comparison performance with your own trained word embeddings

In [53]:
listResX[0]

IndexError: list index out of range

In [55]:
len(listResY)

0